doc/decisions/1_in_discussion/copy_on_write.md

```diff
@@ -327,12 +324,101 @@ API notes:
 - `ksLookup()`:
   - if the `DEL` or `POP` flag is specified, do the COW-stuff as described multiple times now in the operations above
 
-#### Optimizations
+#### Reference Counting
+
+We need reference counting for the internal COW datastructures.
+For the beginning, we can just do it the same way reference counting currently works for `Key` and `KeySet`.
+One tweak though is that the refcount should never be 0, as this does not make sense for internal datastructures.
+
+This means we always increment the refcount after creation and always decrement before deletion, so that the refcount is never zero.
+An example implementation is shown below:
+
+```c
+static void keySetValue(Key * key, void * value, size_t size) {
+  // [...] removal of current value from key
+  struct _KeyData data = keyDataNew (value, size);
+  keyDataIncRef (data);
+  key->data = data;
+}
+
+static void keyDel(Key * key) {
+  keyDataDecRef (key->data);
+  keyDataDel (key->data);
+  // [...] other cleanup
+}
+```
+
+#### Variation 1 - RcBuffer
+
+Instead of using different structs for `_KeyData`, and `_KeyName` use a more generic struct for reference counting.
+This would avoid some duplication on the reference counting code for the key.
+Keysets will still have their own data struct, as it contains more than just a pointer and a size.
+
+```c
+typedef struct {
+	void * data;
+	size_t size;
+	uint16_t refs;
+} RcBuffer;
+
+struct _Key {
+	RcBuffer * uname;
+	RcBuffer * ename; // will be removed soon
+	RcBuffer * value;
+	
+	KeySet * meta;
+	keyflag_t flags;
+	uint16_t refs;
+};
+```
+
+#### Possible Edge Cases
+
+In general, it should be possible to always do copy-on-write.
+From a users perspective, copy-on-write copies of a key (and a keyset) should behave the same.
+There is, however, one edge case: the user modifying the value of a key directly.
+This is shown in the following example:
+
+```c
+Key * key;
+struct foo myFoo = {
+  .x = 0
+};
+keySetBinary (key, &myFoo, sizeof(myFoo));
+
+Key * dup = keyDup (key);
 
-This approach requires more allocations than previously.
-We have not fully benchmarked whether this is a big issue.
-One optimization could be an expanding "pool" of `_KeySetData`, `_KeyData` and `_KeyName`.
-We could then allocate multiple of them at the same time, and borrow and give back instance from and to the pool.
+((struct foo *)keyValue (key))->x = 1;
```

Comment:
This is undefined behavior, the memory returned by `keyValue` must not be modified. This can be simply added as constraint.

Reply:
I don't think it is technically undefined behaviour, but the cast from `const void *` returned by `keyValue` to `struct foo *` is suspicious. However, this code _currently_ works. Therefore, the decision should address it.